### PR TITLE
hinted handoff: improve segment replay logic

### PR DESCRIPTION
**Pull request by piodul**
This series contains two improvements to hint file replay logic in hints manager:

- During replay of a hint file, keeping track of the first hint that fails to be sent is now done via a simple std::optional variable instead of an unordered_set. This slightly reduces complexity of next replay position calculation.
- A corner case is handled: if reading commitlog fails, but there won't be an error related to sending hints, starting position wouldn't be updated. This could cause us to replay more hints than necessary.

Tests:

- unit(dev)
- dtest(hintedhandoff_additional_test, dev)

### CODE REVIEW

**Comment by psarna** [comment]
Looks good. I found it a little confusing that `last_sent_rp` describes the last position that **attempted** to be sent, but I don't have a better suggestion and the commit message explains what happens, so it can stay as is.

**Comment by haaawk** [comment]
I think we should rename _last_not_complete_rp to _first_not_complete_rp

**Comment by haaawk** [comment]
> I think we should rename _last_not_complete_rp to _first_not_complete_rp

or _lowest_not_complete_rp

**Comment by vladzcloudius** [comment]
> I think we should rename _last_not_complete_rp to _first_not_complete_rp

@haaawk 
No, we should not.
Guys, I suggest we have a call and I explain how/why this code works.

File: db/hints/manager.cc [hunk 1]
```diff
@@ -682,6 +682,7 @@ future<> manager::end_point_hints_manager::sender::send_one_mutation(frozen_muta
 }
 
 future<> manager::end_point_hints_manager::sender::send_one_hint(lw_shared_ptr<send_one_file_ctx> ctx_ptr, fragmented_temporary_buffer buf, db::replay_position rp, gc_clock::duration secs_since_file_mod, const sstring& fname) {
+    ctx_ptr->last_sent_rp = rp;
```
Comment (vladzcloudius) [inline db/hints/manager.cc:685]: The name is confusing - it's not "last sent", it's "last attempted".

File: db/hints/manager.hh [hunk 2]
```diff
@@ -104,12 +104,14 @@ public:
                 send_state::segment_replay_failed>>;
 
             struct send_one_file_ctx {
-                send_one_file_ctx(std::unordered_map<table_schema_version, column_mapping>& last_schema_ver_to_column_mapping)
+                send_one_file_ctx(std::unordered_map<table_schema_version, column_mapping>& last_schema_ver_to_column_mapping, db::replay_position last_sent_rp)
                     : schema_ver_to_column_mapping(last_schema_ver_to_column_mapping)
+                    , last_sent_rp(last_sent_rp)
```
Comment (vladzcloudius) [inline db/hints/manager.hh:109]: This is very confusing.
There is not point in trying to save on std::optional here out of a sudden.

There are three distinctive things here:
 1) The position from which we need to restart sending: `_last_not_complete_rp`.
 2) Last attempted send position: `last_sent_rp` (should be `last_attempted_rp`).
 3) First failed position: `first_failed_rp`.

Last two should be optional since you may end up not setting them at all.
And they influence on the value of `_last_not_complete_rp` on the exit from `send_one_file()`.

File: db/hints/manager.hh [hunk 3]
```diff
@@ -97,21 +98,21 @@ public:
...
                 send_one_file_ctx(std::unordered_map<table_schema_version, column_mapping>& last_schema_ver_to_column_mapping)
                     : schema_ver_to_column_mapping(last_schema_ver_to_column_mapping)
                 {}
                 std::unordered_map<table_schema_version, column_mapping>& schema_ver_to_column_mapping;
                 seastar::gate file_send_gate;
-                std::unordered_set<db::replay_position> rps_set; // number of elements in this set is never going to be greater than the maximum send queue length
+                std::optional<db::replay_position> first_failed_rp;
                 send_state_set state;
+
+                void on_hint_send_failure(db::replay_position rp);
```
Comment (vladzcloudius) [inline db/hints/manager.hh:117]: Add `noexcept` qualifier.

File: db/hints/manager.cc [hunk 4]
```diff
@@ -778,10 +779,10 @@ bool manager::end_point_hints_manager::sender::send_one_file(const sstring& fnam
...
     // update the next iteration replay position if needed
     if (ctx_ptr->state.contains(send_state::segment_replay_failed)) {
-        if (ctx_ptr->first_failed_rp) {
-            _last_not_complete_rp = *ctx_ptr->first_failed_rp;
-        }
-
+        // If some hints failed to be sent, first_failed_rp will tell the position of first such hint.
+        // If there was an error thrown by read_log_file function itself, we will retry sending from
+        // the last entry that was successfully read from commitlog (last_sent_rp).
+        _last_not_complete_rp = ctx_ptr->first_failed_rp.value_or(ctx_ptr->last_sent_rp);
```
Comment (vladzcloudius) [inline db/hints/manager.cc:785]: Should be :
```
_last_not_complete_rp = ctx_ptr->first_failed_rp.value_or(ctx_ptr->last_attempted_rp.value_or(_last_not_complete_rp));
```

File: db/hints/manager.cc [hunk 5]
```diff
@@ -724,15 +714,25 @@ future<> manager::end_point_hints_manager::sender::send_one_hint(lw_shared_ptr<s
             } catch (no_column_mapping& e) {
                 manager_logger.debug("send_hints(): {} at {}: {}", fname, rp, e.what());
                 ++this->shard_stats().discarded;
+            } catch (...) {
+                manager_logger.debug("send_hints(): unexpected error in file {} at {}: {}", fname, rp, std::current_exception());
+                ctx_ptr->on_hint_send_failure(rp);
```
[the remark this answers is not in the document]
Comment (piodul) [reply db/hints/manager.cc:720]: I guess the point is to have the "exceptional future ignored" warning appear in the logs, right? Otherwise it should have no effect, as this is a discarded future.

Fixed.

File: db/hints/manager.cc (hunk 1 above)
Comment (piodul) [reply db/hints/manager.cc]: Renamed to "last_attempted_rp"

File: db/hints/manager.hh (hunk 2 above)
Comment (piodul) [reply db/hints/manager.hh]: Yeah, it makes sense to do it in this way. I changed `last_attempted_rp` to be an `std::optional`

File: db/hints/manager.hh (hunk 3 above)
Comment (piodul) [reply db/hints/manager.hh]: Done

File: db/hints/manager.cc (hunk 4 above)
Comment (piodul) [reply db/hints/manager.cc]: Done

**Comment by piodul** [comment]
v2:

- Changed PR description to reflect that this PR is now intended to contain changes from #6401, not just depend on them
- Addressed all current review comments both from this PR and #6401

File: db/hints/manager.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (piodul) [reply db/hints/manager.cc:720]: After thinking about it some more, I realized that we should be more careful here. If we encounter an unknown exception, we don't know if it is caused by temporary failure or not. 

This line, in its current form, will cause us to fail send_one_file in case of unknown exception. This could cause hints manager to get stuck if the error caused by the hint is persistent.

I'd vote for dropping in this case instead (and will pre-emptively change this PR to do that instead). The worst that could happen is that we lose some hints due to an unknown error that could go away after retry.

**Comment by piodul** [comment]
v3:
- Rebased onto recent master
- Drop hint instead of failing send_one_file when an unknown error is encountered

File: db/hints/manager.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (vladzcloudius) [reply db/hints/manager.cc]: @piodul No, for the default `catch` clause  you want to treat it as an error and in this case we return exceptional future.

[the remark this answers is not in the document]
Comment (vladzcloudius) [reply db/hints/manager.cc]: > I guess the point is to have the "exceptional future ignored" warning appear in the logs, right? Otherwise it should have no effect, as this is a discarded future.
> 
@piodul No, no need for any extra warning. 
For the default `catch` clause  you want to treat it as an error and in this case we return exceptional future.

[the remark this answers is not in the document]
Comment (vladzcloudius) [reply db/hints/manager.cc]: > I realized that we should be more careful here. If we encounter an unknown exception, we don't know if it is caused by temporary failure or not.
> 
> (The issue mentioned in this comment is not related to returning exceptional future, but is related to this particular piece of code)
> 
> This line, in its current form, will cause us to fail send_one_file in case of unknown exception. This could cause hints manager to get stuck if the error caused by the hint is persistent.
> 
> I'll change it so that we drop the hint instead. It is what we did before this patch, and it is safer this way.

@piodul I'm not sure I can agree with you that before that we are discarding the hint that was handled when we hit the default `catch(...)` clasue today.
As far as I can tell we would fail in sending the current file but the rp for that hint would remain in the rp_set. Therefore would would re-send in the next file_send attempt.

The reason we want to return an exceptional future here is to break sending this file sooner.

The reason for the failure could be any kind of exception in the scope in question, e.g. in `clock::now()`.

File: db/hints/manager.cc [hunk 6]
```diff
@@ -725,15 +714,26 @@ future<> manager::end_point_hints_manager::sender::send_one_hint(lw_shared_ptr<s
             } catch (no_column_mapping& e) {
                 manager_logger.debug("send_hints(): {} at {}: {}", fname, rp, e.what());
                 ++this->shard_stats().discarded;
+            } catch (...) {
+                manager_logger.debug("send_hints(): unexpected error in file {} at {}: {}", fname, rp, std::current_exception());
+                ++this->shard_stats().discarded;
```
Comment (vladzcloudius) [inline db/hints/manager.cc:720]: Please, don't discard this hint. 
The v2 version was better: 
 1) Call `ctx_ptr->on_hint_send_failure(rp);`
 2) Add a debug message.

And then simply add `return make_exception_future<>(std::current_exception());`

File: db/hints/manager.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (piodul) [reply db/hints/manager.cc]: @vladzcloudius Returning an exceptional future there will not have any more effect than a big warning with a backtrace, printed by the reactor during runtime ("exceptional future ignored"). [Here](https://github.com/scylladb/scylla/blob/517f2c049058a2231853515b43732fadba77017a/db/hints/manager.cc#L687) you can observe that nobody waits for that future - it is explicitly discarded.

In current implementation, no exception handler will be called for that error. Hint position will remain in rps_set, and `segment_replay_failed` won't be set.

It might be confusing because sender::send_one_hint has an exception handler at the top level of `send_one_hint`, but unhandled exceptions in the discarded future won't reach it.

[the remark this answers is not in the document]
Comment (piodul) [reply db/hints/manager.cc]: > @piodul I'm not sure I can agree with you that we are discarding the hint that was handled when we hit the default catch(...) clasue today.
> As far as I can tell we would fail in sending the current file but the rp for that hint would remain in the rp_set. Therefore we would re-send it in the next file_send attempt.

@vladzcloudius See my comment above for the explanation why (I think) current logic quietly discards the hint in this case.

> The reason for the failure could be any kind of exception in the scope in question, e.g. in clock::now().

My argument for dropping the hint is that in case of `catch (...)` we don't know if the exception will continue to be thrown every time we try to process the same hint. An example would be an exception caused by an incorrectly encoded mutation. Such hint would prevent us from finishing the commitlog segment that contains it.

I think that we should try to add more `catch(some_exception& e)` blocks and choose for each case if we should drop the hint or retry sending the file (not sure if it is in scope of this patch). For errors that we didn't account for, I think it will be safer to drop the hint so that we won't accidentally make hint manager get stuck in a loop, trying to send hints from the same file again and again.

[the remark this answers is not in the document]
Comment (vladzcloudius) [reply db/hints/manager.cc]: > @vladzcloudius Returning an exceptional future there will not have any more effect than a big warning with a backtrace, printed by the reactor during runtime ("exceptional future ignored"). [Here](https://github.com/scylladb/scylla/blob/517f2c049058a2231853515b43732fadba77017a/db/hints/manager.cc#L687) you can observe that nobody waits for that future - it is explicitly discarded.
> 
> In current implementation, no exception handler will be called for that error. Hint position will remain in rps_set, and `segment_replay_failed` won't be set.
> 
> It might be confusing because sender::send_one_hint has an exception handler at the top level of `send_one_hint`, but unhandled exceptions in the discarded future won't reach it.

@piodul I must disagree: returning exceptional future is not going to be ignored since it's going to be eventually caught and handled here: https://github.com/scylladb/scylla/blob/master/db/hints/manager.cc#L732

That's the intent and let's preserve this behaviour. 

I also agree that if there are other special cases like malformed mutation they should be explicitly handled. 

As for the malformed mutation AFAIU this should never happen - in this case I'd expect the corresponding hint segment checksum check to fail which in line should cause the read_log_file() to throw `db::commitlog::segment_error` which is caught and handled in `sender::send_one_file()`.

**Comment by avikivity** [comment]
@psarna please merge if it passed your review and you don't need additional reviewers.

**Comment by piodul** [comment]
@avikivity @psarna Please do not merge yet. I'm planning to send a small update to this PR today and would like to let Vlad take a look at it before merging.

**Comment by piodul** [comment]
v3: rebased and updated. Now, if an unexpected exception occurs while hint is being sent, it won't cause the hint to be dropped, instead it will be retried later.

Diff with v2:

```
diff --git a/db/hints/manager.cc b/db/hints/manager.cc
index 2bd491c23..63c41a832 100644
--- a/db/hints/manager.cc
+++ b/db/hints/manager.cc
@@ -717,8 +717,7 @@ future<> manager::end_point_hints_manager::sender::send_one_hint(lw_shared_ptr<s
                 ++this->shard_stats().discarded;
             } catch (...) {
                 manager_logger.debug("send_hints(): unexpected error in file {} at {}: {}", fname, rp, std::current_exception());
-                ++this->shard_stats().discarded;
-                return make_exception_future<>(std::current_exception());
+                ctx_ptr->on_hint_send_failure(rp);
             }
             return make_ready_future<>();
         }).finally([units = std::move(units), ctx_ptr] {});
```

File: db/hints/manager.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (piodul) [reply db/hints/manager.cc:720]: @vladzcloudius Sorry for the long delay. I updated the PR so that in case of an unexpected exception, the hint won't be dropped, but retried later.

> @piodul I must disagree: returning exceptional future is not going to be ignored since it's going to be eventually caught and handled here: https://github.com/scylladb/scylla/blob/master/db/hints/manager.cc#L732

As for the exceptional future issue... the exception handler you linked won't handle it. The hint sending task starts [here](https://github.com/scylladb/scylla/blob/4e7929609086e99afd17799ec4eaa641d871f659/db/hints/manager.cc#L687), and ends [here](https://github.com/scylladb/scylla/blob/4e7929609086e99afd17799ec4eaa641d871f659/db/hints/manager.cc#L731). It's basically:

```C++
(void)with_gate(_file_send_gate, [] { /* stuff */ }).finally([] { /* stuff */ });
```

`with_gate` calls its lambda argument using `futurize_invoke`, so unless the gate is closed, it doesn't throw. A C++ exception thrown inside the lambda argument will be converted into exceptional future. Because the result is discarded with `(void)`, it will be just reported as an exceptional future.

Therefore there is no need for returning `make_exception_future`. A message will be printed similar to this: "send_hints(): unexpected error in file {} at {}: {}", and sending of the hint will be retried later.

**Comment by piodul** [comment]
@avikivity We have dtests for hinted handoff here: https://github.com/scylladb/scylla-dtest/blob/next/hintedhandoff_additional_test.py

We also have a SCT nemesis: https://github.com/scylladb/scylla-cluster-tests/blob/master/sdcm/nemesis.py#L1309

AFAIK there are no Scylla unit tests for HH.

Our dtests verify that HH works in idealized scenarios, where reading from commitlog and sending hints is expected to succeed. I think the closest we have to testing failure scenarios is the SCT nemesis.

We can extend hinted handoff to have error injection, and then write some dtests for failure cases. I created an issue for adding error injections: https://github.com/scylladb/scylla/issues/6649

**Comment by vladzcloudius** [comment]
> v3: rebased and updated. Now, if an unexpected exception occurs while hint is being sent, it won't cause the hint to be dropped, instead it will be retried later.
> 
> Diff with v2:
> 
> ```
> diff --git a/db/hints/manager.cc b/db/hints/manager.cc
> index 2bd491c23..63c41a832 100644
> --- a/db/hints/manager.cc
> +++ b/db/hints/manager.cc
> @@ -717,8 +717,7 @@ future<> manager::end_point_hints_manager::sender::send_one_hint(lw_shared_ptr<s
>                  ++this->shard_stats().discarded;
>              } catch (...) {
>                  manager_logger.debug("send_hints(): unexpected error in file {} at {}: {}", fname, rp, std::current_exception());
> -                ++this->shard_stats().discarded;
> -                return make_exception_future<>(std::current_exception());
> +                ctx_ptr->on_hint_send_failure(rp);
>              }
>              return make_ready_future<>();
>          }).finally([units = std::move(units), ctx_ptr] {});
> ```

@piodul The logic above was intentional - if an unexpected exception occurred there is no guarantee that it will not happen again the next time you replay the same hint. In this case we'll spin on that hint forever.

The current logic is meant to prevent this.

**Comment by piodul** [comment]
@psarna ping, got approval from Vlad and Avi

**Comment by psarna** [comment]
3bcc2e8f0916324696661dd07f29e4c085ce75e6
